subsys/bluetooth/audio/has.c

```diff
@@ -0,0 +1,1330 @@
+	}
+
+	req = net_buf_simple_pull_mem(buf, sizeof(*req));
+
+	return preset_name_set(req->idx, req->name, buf->len);
+}
+#else
+	return -EOPNOTSUPP;
+}
+#endif /* CONFIG_BT_HAS_PRESET_NAME_DYNAMIC */
+
+static int call_ops_set_active_preset(uint8_t id, bool sync)
+{
+	if (preset_ops == NULL) {
+		return BT_GATT_ERR(BT_HAS_ERR_OPERATION_NOT_POSSIBLE);
+	}
+
+	if (preset_ops->active_set(&has_local, id, sync) != 0) {
+		return BT_GATT_ERR(BT_HAS_ERR_OPERATION_NOT_POSSIBLE);
+	}
+
+	return 0;
+}
+
+static int handle_set_active_preset(struct has_client *client, struct net_buf_simple *buf,
+				    bool sync)
+{
+	const struct bt_has_cp_set_active_preset_req *req;
+	struct has_preset *preset;
+
+	if (buf->len < sizeof(*req)) {
+		return BT_GATT_ERR(BT_HAS_ERR_INVALID_PARAM_LEN);
+	}
+
+	req = net_buf_simple_pull_mem(buf, sizeof(*req));
+
+	preset = preset_get(req->idx);
+	if (!preset) {
+		return BT_GATT_ERR(BT_ATT_ERR_OUT_OF_RANGE);
+	}
+
+	if (!(preset->properties & BT_HAS_PROP_AVAILABLE)) {
+		return BT_GATT_ERR(BT_HAS_ERR_OPERATION_NOT_POSSIBLE);
+	}
+
+	return call_ops_set_active_preset(preset->id, sync);
+}
+
+static bool find_available(struct has_preset *preset, void *user_data)
+{
+	struct has_preset **found = user_data;
+
+	if ((preset->properties & BT_HAS_PROP_AVAILABLE) && !preset->hidden) {
+		*found = preset;
+		return true;
+	}
+
+	return false;
+}
+
+static int handle_set_next_preset(struct has_client *client, struct net_buf_simple *buf, bool sync)
+{
+	struct has_preset *preset = NULL;
+
+	preset_foreach(has_local.active_id + 1, last_preset_id, 1, find_available, &preset);
+	if (preset) {
+		return call_ops_set_active_preset(preset->id, sync);
+	}
+
+	preset_foreach(0x01, has_local.active_id - 1, 1, find_available, &preset);
+	if (preset) {
+		return call_ops_set_active_preset(preset->id, sync);
+	}
+
+	return BT_GATT_ERR(BT_HAS_ERR_OPERATION_NOT_POSSIBLE);
+}
+
+static int handle_set_prev_preset(struct has_client *client, struct net_buf_simple *buf, bool sync)
+{
+	struct has_preset *preset = NULL;
+
+	preset_foreach(0x01, has_local.active_id - 1, 1, find_available, &preset);
+	if (preset) {
+		return call_ops_set_active_preset(preset->id, sync);
+	}
+
+	preset_foreach(has_local.active_id + 1, last_preset_id, 1, find_available, &preset);
+	if (preset) {
+		return call_ops_set_active_preset(preset->id, sync);
+	}
+
+	return BT_GATT_ERR(BT_HAS_ERR_OPERATION_NOT_POSSIBLE);
+}
+
+static ssize_t control_point_rx(struct bt_conn *conn, const struct bt_gatt_attr *attr,
+				const void *data, uint16_t len, uint16_t offset, uint8_t flags)
+{
+	const struct bt_has_cp_hdr *hdr;
+	struct has_client *client;
+	struct net_buf_simple buf;
+	ssize_t ret;
+
+	client = client_find(conn);
+	if (!client) {
+		return BT_GATT_ERR(BT_ATT_ERR_UNLIKELY);
+	}
+
+	if (offset > 0) {
+		return BT_GATT_ERR(BT_ATT_ERR_INVALID_OFFSET);
+	}
+
+	if (len < sizeof(*hdr)) {
+		return BT_GATT_ERR(BT_ATT_ERR_INVALID_ATTRIBUTE_LEN);
+	}
```

Comment:
Consider moving these checks up before `client_find` - No need to find a client in those cases